demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+        }
+    }
+
+    return ( int ) res;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV generateKeyPairEC( CK_SESSION_HANDLE session,
+                                uint8_t * privateKeyLabel,
+                                uint8_t * publicKeyLabel,
+                                CK_OBJECT_HANDLE_PTR privateKeyHandlePtr,
+                                CK_OBJECT_HANDLE_PTR publicKeyHandlePtr )
+{
+    CK_RV result;
+    CK_MECHANISM mechanism = { CKM_EC_KEY_PAIR_GEN, NULL_PTR, 0 };
+    CK_FUNCTION_LIST_PTR functionList;
+    CK_BYTE ecParams[] = pkcs11DER_ENCODED_OID_P256; /* prime256v1 */
+    CK_KEY_TYPE keyType = CKK_EC;
+
+    CK_BBOOL trueObject = CK_TRUE;
+    CK_ATTRIBUTE publicKeyTemplate[] =
+    {
+        { CKA_KEY_TYPE,  NULL /* &keyType */,    sizeof( keyType )                         },
+        { CKA_VERIFY,    NULL /* &trueObject */, sizeof( trueObject )                      },
+        { CKA_EC_PARAMS, NULL /* ecParams */,    sizeof( ecParams )                        },
+        { CKA_LABEL,     publicKeyLabel,         strlen( ( const char * ) publicKeyLabel ) }
+    };
+
+    /* Aggregate initializers must not use the address of an automatic variable. */
+    publicKeyTemplate[ 0 ].pValue = &keyType;
+    publicKeyTemplate[ 1 ].pValue = &trueObject;
+    publicKeyTemplate[ 2 ].pValue = &ecParams;
+
+    CK_ATTRIBUTE privateKeyTemplate[] =
+    {
+        { CKA_KEY_TYPE, NULL /* &keyType */,    sizeof( keyType )                          },
+        { CKA_TOKEN,    NULL /* &trueObject */, sizeof( trueObject )                       },
+        { CKA_PRIVATE,  NULL /* &trueObject */, sizeof( trueObject )                       },
+        { CKA_SIGN,     NULL /* &trueObject */, sizeof( trueObject )                       },
+        { CKA_LABEL,    privateKeyLabel,        strlen( ( const char * ) privateKeyLabel ) }
+    };
+
+    /* Aggregate initializers must not use the address of an automatic variable. */
+    privateKeyTemplate[ 0 ].pValue = &keyType;
+    privateKeyTemplate[ 1 ].pValue = &trueObject;
+    privateKeyTemplate[ 2 ].pValue = &trueObject;
+    privateKeyTemplate[ 3 ].pValue = &trueObject;
+
+    result = C_GetFunctionList( &functionList );
+
+    result = functionList->C_GenerateKeyPair( session,
+                                              &mechanism,
+                                              publicKeyTemplate,
+                                              sizeof( publicKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                              privateKeyTemplate, sizeof( privateKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                              publicKeyHandlePtr,
+                                              privateKeyHandlePtr );
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+bool generateKeyAndCsr( CK_SESSION_HANDLE p11Session,
+                        char * pCsrBuffer,
```

Comment:
Would be better to pass the labels explictly to this utility function to make it easier for customers to see labels being used within the demo file